### DISCUSSION TITLE
Publishing a MediatR event from a MediatR command handler

### DISCUSSION THREAD

**Discussion by RJR-Robert-Kewei-Zhang** [LuckyPennySoftware/MediatR#755] [Q&A]
Hi @jbogard,  

I need to publish a MediatR event and I am wondering if I can invoke the publish method from a MetiatR command handler. 

Thanks,
Robert

**Comment by jbogard** [comment]
Yes you can. Just keep in mind the notification handlers run in-process with the command handler.

**Reply by RJR-Robert-Kewei-Zhang** [reply]
Thank you Jimmy, it is really helpful
Robert

**Reply by RJR-Robert-Kewei-Zhang** [reply]
Hi Jimmy,

Below is the code sample of how I am using it, just want to be sure that we are on the same page

//MediatR, invoking send method
using MediatR;
using mediatrAPI.Commands;

namespace mediatrAPI.Controllers
{
    [ApiController]
    [Route("[controller]")]
    public class OrderController : ControllerBase
    {
        private readonly IMediator _mediator;

        public OrderController(IMediator mediator)
        {
            _mediator = mediator;

        }

        [HttpPost]
        [Route("PlaceOrder")]
        public Task PostAsync([FromBody] PlaceOrderCommand order) =>
              _mediator.Send(order);
    }
}

//MediatR, command handler
using MediatR;
using mediatrAPI.Events;

namespace mediatrAPI.Commands
{
    public class PlaceOrderCommand : IRequest
    {
        public string CustomerName { get; set; }
        public OrderLine[] OrderLines { get; set; }
        public class OrderLine
        {
            public string ProductName { get; set; }
            public int Quantity { get; set; }
        }

        public class PlaceOrderRequestHandler : IRequestHandler<PlaceOrderCommand>
        {
            private readonly IMediator _mediator;
            public PlaceOrderRequestHandler(IMediator mediator)
            {
                _mediator = mediator;
            }
            
            
            public async Task<Unit> Handle(PlaceOrderCommand request, CancellationToken cancellationToken)
            {
                try
                {
                    //Some code to save the order infomration
                    //......
                    //

                    //Invoking the mediator publish method
                    await _mediator.Publish(new OrderPlacedEvent
                    {
                        productName = "Garden tools",
                        numberOfProduct = 2
                    }, cancellationToken);
                    return Unit.Value;
                }
                catch (Exception e)
                {
                    Console.WriteLine(DateTime.Now.ToString() + ": " + e.Message);
                    return Unit.Value;
                }
            }
        }
    }
}

Thanks,
Robert
